File: bindings/ruby/lib/pqc_dyber/kem.rb
# libpqc-dyber - Post-Quantum Cryptography Library
# Copyright (c) 2024-2026 Dyber, Inc.
# SPDX-License-Identifier: Apache-2.0 OR MIT
#
# KEM (Key Encapsulation Mechanism) class.

module PqcDyber
  class KEM
    attr_reader :algorithm

    # Create a KEM instance for the given algorithm.
    #
    # @param algorithm [String] e.g. "ML-KEM-768"
    # @raise [RuntimeError] if algorithm is not supported
    def initialize(algorithm)
      @algorithm = algorithm
      @handle = PqcDyber.pqc_kem_new(algorithm)
      raise "Unsupported KEM algorithm: #{algorithm}" if @handle.null?

      ObjectSpace.define_finalizer(self, self.class._release(@handle))
    end

    def public_key_size
      PqcDyber.pqc_kem_public_key_size(@handle)
    end

    def secret_key_size
      PqcDyber.pqc_kem_secret_key_size(@handle)
    end

    def ciphertext_size
      PqcDyber.pqc_kem_ciphertext_size(@handle)
    end

    def shared_secret_size
      PqcDyber.pqc_kem_shared_secret_size(@handle)
    end

    def security_level
      PqcDyber.pqc_kem_security_level(@handle)
    end

    # Generate a keypair.
    #
    # @return [Hash] { public_key: String, secret_key: String }
    def keygen
      pk = FFI::MemoryPointer.new(:uint8, public_key_size)
      sk = FFI::MemoryPointer.new(:uint8, secret_key_size)
      rc = PqcDyber.pqc_kem_keygen(@handle, pk, sk)
      raise "KEM keygen failed: #{PqcDyber.pqc_status_string(rc)}" unless rc == 0

      { public_key: pk.read_bytes(public_key_size),
        secret_key: sk.read_bytes(secret_key_size) }
    end

    # Encapsulate: generate shared secret and ciphertext from a public key.
    #
    # @param public_key [String]
    # @return [Hash] { ciphertext: String, shared_secret: String }
    def encaps(public_key)
      ct = FFI::MemoryPointer.new(:uint8, ciphertext_size)
      ss = FFI::MemoryPointer.new(:uint8, shared_secret_size)
      pk_buf = FFI::MemoryPointer.from_string(public_key)

      rc = PqcDyber.pqc_kem_encaps(@handle, ct, ss, pk_buf)
      raise "KEM encaps failed: #{PqcDyber.pqc_status_string(rc)}" unless rc == 0

      { ciphertext: ct.read_bytes(ciphertext_size),
        shared_secret: ss.read_bytes(shared_secret_size) }
    end

    # Decapsulate: recover shared secret from ciphertext using secret key.
    #
    # @param ciphertext [String]
    # @param secret_key [String]
    # @return [String] the shared secret
    def decaps(ciphertext, secret_key)
      ss = FFI::MemoryPointer.new(:uint8, shared_secret_size)
      ct_buf = FFI::MemoryPointer.from_string(ciphertext)
      sk_buf = FFI::MemoryPointer.from_string(secret_key)

      rc = PqcDyber.pqc_kem_decaps(@handle, ss, ct_buf, sk_buf)
      raise "KEM decaps failed: #{PqcDyber.pqc_status_string(rc)}" unless rc == 0

      ss.read_bytes(shared_secret_size)
    end

    # @private
    def self._release(handle)
      proc { PqcDyber.pqc_kem_free(handle) }
    end
  end
end

File: bindings/swift/Sources/PQCDyber/KEM.swift
// libpqc-dyber - Post-Quantum Cryptography Library
// Copyright (c) 2024-2026 Dyber, Inc.
// SPDX-License-Identifier: Apache-2.0 OR MIT

import Foundation
import Cpqc

/// Key Encapsulation Mechanism (KEM) operations.
public final class KEM {

    /// KEM keypair result.
    public struct KeyPair {
        public let publicKey: Data
        public let secretKey: Data
    }

    /// KEM encapsulation result.
    public struct EncapsResult {
        public let ciphertext: Data
        public let sharedSecret: Data
    }

    private let handle: OpaquePointer

    /// Create a KEM context for the specified algorithm.
    /// - Parameter algorithm: Algorithm name, e.g. "ML-KEM-768".
    /// - Throws: `PQCError.unsupportedAlgorithm` if the algorithm is not available.
    public init(algorithm: String) throws {
        guard let h = pqc_kem_new(algorithm) else {
            throw PQCError.unsupportedAlgorithm(algorithm)
        }
        self.handle = h
    }

    deinit {
        pqc_kem_free(handle)
    }

    /// The algorithm name.
    public var algorithm: String {
        String(cString: pqc_kem_algorithm(handle))
    }

    /// Public key size in bytes.
    public var publicKeySize: Int {
        pqc_kem_public_key_size(handle)
    }

    /// Secret key size in bytes.
    public var secretKeySize: Int {
        pqc_kem_secret_key_size(handle)
    }

    /// Ciphertext size in bytes.
    public var ciphertextSize: Int {
        pqc_kem_ciphertext_size(handle)
    }

    /// Shared secret size in bytes.
    public var sharedSecretSize: Int {
        pqc_kem_shared_secret_size(handle)
    }

    /// NIST security level (1-5).
    public var securityLevel: Int {
        Int(pqc_kem_security_level(handle))
    }

    /// Generate a new keypair.
    /// - Returns: A `KeyPair` containing the public and secret keys.
    /// - Throws: `PQCError.operationFailed` on failure.
    public func keygen() throws -> KeyPair {
        var pk = Data(count: publicKeySize)
        var sk = Data(count: secretKeySize)

        let rc = pk.withUnsafeMutableBytes { pkPtr in
            sk.withUnsafeMutableBytes { skPtr in
                pqc_kem_keygen(handle,
                               pkPtr.bindMemory(to: UInt8.self).baseAddress!,
                               skPtr.bindMemory(to: UInt8.self).baseAddress!)
            }
        }
        guard rc == 0 else { throw PQCError.operationFailed(rc) }
        return KeyPair(publicKey: pk, secretKey: sk)
    }

    /// Encapsulate: generate shared secret and ciphertext from a public key.
    /// - Parameter publicKey: The recipient's public key.
    /// - Returns: An `EncapsResult` with ciphertext and shared secret.
    /// - Throws: `PQCError.operationFailed` on failure.
    public func encaps(publicKey: Data) throws -> EncapsResult {
        var ct = Data(count: ciphertextSize)
        var ss = Data(count: sharedSecretSize)

        let rc = ct.withUnsafeMutableBytes { ctPtr in
            ss.withUnsafeMutableBytes { ssPtr in
                publicKey.withUnsafeBytes { pkPtr in
                    pqc_kem_encaps(handle,
                                   ctPtr.bindMemory(to: UInt8.self).baseAddress!,
                                   ssPtr.bindMemory(to: UInt8.self).baseAddress!,
                                   pkPtr.bindMemory(to: UInt8.self).baseAddress!)
                }
            }
        }
        guard rc == 0 else { throw PQCError.operationFailed(rc) }
        return EncapsResult(ciphertext: ct, sharedSecret: ss)
    }

    /// Decapsulate: recover shared secret from ciphertext using secret key.
    /// - Parameters:
    ///   - ciphertext: The ciphertext from encapsulation.
    ///   - secretKey: The recipient's secret key.
    /// - Returns: The shared secret as `Data`.
    /// - Throws: `PQCError.operationFailed` on failure.
    public func decaps(ciphertext: Data, secretKey: Data) throws -> Data {
        var ss = Data(count: sharedSecretSize)

        let rc = ss.withUnsafeMutableBytes { ssPtr in
            ciphertext.withUnsafeBytes { ctPtr in
                secretKey.withUnsafeBytes { skPtr in
                    pqc_kem_decaps(handle,
                                   ssPtr.bindMemory(to: UInt8.self).baseAddress!,
                                   ctPtr.bindMemory(to: UInt8.self).baseAddress!,
                                   skPtr.bindMemory(to: UInt8.self).baseAddress!)
                }
            }
        }
        guard rc == 0 else { throw PQCError.operationFailed(rc) }
        return ss
    }
}

/// Errors from PQC operations.
public enum PQCError: Error, LocalizedError {
    case unsupportedAlgorithm(String)
    case operationFailed(Int32)

    public var errorDescription: String? {
        switch self {
        case .unsupportedAlgorithm(let name):
            return "Unsupported algorithm: \(name)"
        case .operationFailed(let code):
            if let ptr = pqc_status_string(code) {
                return String(cString: ptr)
            }
            return "PQC operation failed with code \(code)"
        }
    }
}
